### ISSUE TITLE
C++ lambda demangling

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#380]
Please see the below example:
```cpp
int main()
{
    auto f = [](int a, int b) { return a + b; };

    return f(1, 2);
}
```
It shows replay output as follows:
```
$ g++ -std=c++14 -pg lambda.cpp
$ uftrace a.out
# DURATION     TID     FUNCTION
   1.053 us [116916] | __monstartup();
   1.310 us [116916] | __cxa_atexit();
            [116916] | main() {
   0.263 us [116916] |   main::operator ();
   1.236 us [116916] | } /* main */
```
I think `main::operator()` doesn't show it's a lambda function.

But if full demangler is used, it shows more meaningful symbol name:
```
$ uftrace --demangle=full a.out
# DURATION     TID     FUNCTION
   1.050 us [116892] | __monstartup();
   1.016 us [116892] | __cxa_atexit();
            [116892] | main() {
   0.184 us [116892] |   main::{lambda(int, int)#1}::operator()(int, int) const();
   0.880 us [116892] | } /* main */
```
Can it be `main::{lambda(int, int)#1}::operator()` instead of `main::operator()` to explicitly show its a lambda function call?

**Comment by honggyukim** [comment]
clang shows its name more like lambda function so doesn't have to be changed.
```
$ clang++ -std=c++14 -pg lambda.cpp

$ uftrace a.out
# DURATION     TID     FUNCTION
            [117154] | main() {
   0.267 us [117154] |   main::$_0::operator ();
   1.710 us [117154] | } /* main */

$ uftrace --demangle=full a.out
# DURATION     TID     FUNCTION
            [117134] | main() {
   0.270 us [117134] |   main::$_0::operator()(int, int) const();
   1.747 us [117134] | } /* main */
```

**Comment by namhyung** [comment]
Yeah, I'd like to follow clang's naming as it's shorter and clear.

**Comment by namhyung** [comment]
Oops, sorry.  I don't know how exactly it works.. How about this?

```diff
diff --git a/utils/demangle.c b/utils/demangle.c
index e3fbcdfb..5c174983 100644
--- a/utils/demangle.c
+++ b/utils/demangle.c
@@ -1267,7 +1267,7 @@ static int dd_unqualified_name(struct demangle_data *dd)
                        dd->type--;
                }
                else if (c1 == 'l') {
-                       int n = 0;
+                       int n = -1;
                        char buf[32];

                        /* closure type name (or lambda) */
@@ -1294,7 +1294,7 @@ static int dd_unqualified_name(struct demangle_data *dd)
                        if (dd->newpos)
                                dd_append(dd, "::");

-                       snprintf(buf, sizeof(buf), "$_%d", n);
+                       snprintf(buf, sizeof(buf), "$_%d", n + 1);
                        dd_append(dd, buf);
                }
                else {
```

**Comment by honggyukim** [comment]
It works fine for g++ compile binary. Thanks : )
```
$ uftrace a.out
# DURATION     TID     FUNCTION
   1.240 us [133156] | __monstartup();
   1.010 us [133156] | __cxa_atexit();
            [133156] | main() {
   0.180 us [133156] |   main::$_0::operator();
   0.171 us [133156] |   main::$_1::operator();
   1.347 us [133156] | } /* main */
```

**Comment by honggyukim** [comment]
FYI, if demangle full is used with dwarf info, it looks as follows:
```
$ g++ -std=c++14 -pg -g test.cc
$ uftrace -A. -R. --demangle=full a.out
# DURATION     TID     FUNCTION
   1.440 us [134029] | __monstartup();
   1.074 us [134029] | __cxa_atexit();
            [134029] | main() {
   0.227 us [134029] |   main::{lambda(int, int)#1}::operator()(int, int) const();
   0.170 us [134029] |   main::{lambda(int, int, int)#2}::operator()(int, int, int) const();
   1.740 us [134029] | } = 9; /* main */
```
But clang compiled binary shows args/retval correctly.
```
$ clang++ -std=c++14 -pg -g test.cc
$ uftrace -A. -R. --demangle=full a.out
# DURATION     TID     FUNCTION
            [134003] | main() {
   1.607 us [134003] |   main::$_0::operator()(int, int) const(0x7ffea525c3d8, 1, 2) = 3;
   0.530 us [134003] |   main::$_1::operator()(int, int, int) const(0x7ffea525c3d0, 1, 2, 3) = 6;
   6.137 us [134003] | } = 9; /* main */
```

**Comment by namhyung** [comment]
Strange.. I will take a look

**Comment by namhyung** [comment]
Merged: 02fdc618
